src/main/java/org/jmxtrans/agent/Query.java

```diff
@@ -131,20 +131,35 @@ public void collectAndExport(@Nonnull MBeanServer mbeanServer, @Nonnull OutputWr
 
         for (ObjectName on : objectNames) {
             try {
-                if (attribute == null || attribute.isEmpty()) {
-                    for (MBeanAttributeInfo mBeanAttributeInfo : mbeanServer.getMBeanInfo(on).getAttributes()) {
-                        collectAndExportAttribute(mbeanServer, outputWriter, on, mBeanAttributeInfo.getName());
-                    }
-
-                } else {
-                    collectAndExportAttribute(mbeanServer, outputWriter, on, attribute);
-                }
+                collectAndExportForObjectName(mbeanServer, on, outputWriter);
             } catch (Exception e) {
                 logger.log(Level.WARNING, "Exception collecting " + on + "#" + attribute + (key == null ? "" : "#" + key), e);
             }
         }
     }
 
+    private void collectAndExportForObjectName(MBeanServer mbeanServer, ObjectName on, OutputWriter outputWriter)
+            throws Exception {
+        for (String attribute : resolveAttributes(mbeanServer, on)) {
```

Comment:
I meant caching the result of `split`.

Could we replace the field `protected final String attribute` by `protected final List<String> attributes`?

Do you see a scenario in which that would cause issues?